GPU device placement issues

LouisRouillard opened the pull request:
I think most of the issues encountered by users result from `device` strings processing and mismatch. I'm trying a bunch of error-prone combinations and trying to see how can those mismatch be catched and corrected.

For now I only tried very small changes.

2 (known) issues remain:
- prior device mismatch
- embedding net with custom density estimator mismatch

But both items do not expose any device attribute for now so maybe more invasive changes could be needed

janfb commented:
looks good, thanks! 

Regarding the prior device: there is a function for checking the prior device, but it is not used currently: 
https://github.com/mackelab/sbi/blob/bc4d43bf60ec714790ef8baa4b08cc6e5b821e45/sbi/utils/torchutils.py#L42-L51

We could use this function in `sbi/inference/base.py` to check the prior (if it is not `None`). 

Regarding the embedding and density estimator: this happens outside of the SBI loop, no? E.g., the user would use `get_nn_models` function to build their density estimator with embedding net, right? Here one quick fix would be to check the embedding net for its device, and then 
a) throw an error that it should be on the CPU for now, or
b) throw a warning and move it to CPU to then compose it with the density estimator on the CPU. 

I would tend to b). And you? 

@michaeldeistler @jan-matthis what do you think?

michaeldeistler commented:
I'm having trouble understanding which cases are going wrong when.

Is the issue that `get_posterior_nn()` (etc) can not handle the `embedding_net` being on the GPU?

janfb commented:
Yes, that's one of the problems. We designed `sbi` such that the user would just say `device="cuda"` and we take care of moving everything to the device. However, it happens that users have their priors, data, or embeddings on the device already, and then `posterior_nn()` would compose a Flow with mixed devices and things break.

LouisRouillard commented:
Hey ! Sorry I had a bunch of stuff to work on, I'll work on the aforementioned issues some more tonight. @janfb I personally like APIs that are not error compliant but indicate clearly what went wrong -so option **a**-. But `sbi` is generally built around the "warn and correct" concept so I'll try to implement **b**

LouisRouillard commented:
@michaeldeistler I can show you the cases that go wrong in screenshare like I did with @janfb at some point if you want?

LouisRouillard commented:
I've done a couple functions that check what seems to be the most common error cases. I'm not sure I have propagated those functions everywhere needed though. For the `embedding_net` I managed to implement a "warn and correct" behavior, but not for the prior check because I feel like I would have needed to "wrap" a potentially custom prior into another prior that replicates the sampling and then moves the data accordingly. Tell me what you think ?

LouisRouillard commented:
Arf I realize I redefined a prior checking function. I'll remove it.

michaeldeistler commented:
@LouisRouillard I think quickest for now would be to just do a single commit with the black formatting changes. Then I'll quickly have a look :)

janfb commented on sbi/utils/user_input_checks.py:
```diff
@@ -372,6 +372,23 @@ def check_prior_support(prior):
 
 
+def check_embedding_net_device(embedding_net: nn.Module, batch_y: torch.Tensor) -> None:
+    batch_y_device = batch_y.device
```
what if the user passes data on GPU? Then this check would pass and the embedding would remain on the GPU and cause issues with the density estimator that is constructed on CPU, no? 

If we assume that we always build the density estimator on the CPU, then we could just always move the embedding to the GPU, no? (and warn, or throw an error depending on whether we want to `warn and correct` or not. )

LouisRouillard replied:
I'm not sure this situation will arise in the sense that the mismatch between `prior` and `estimator` is checked before that step? So before this is ran we know that `batch_y` (generated from the `prior`) is the same device as `estimator`, and all that remains to do is to check that the `embedder` is also in that case? I guess the question is: is there  use case -unknown to me- where the `build` functions are used outside of the `classic` "pipeline" ?

LouisRouillard commented:
Ok so I adaanced on the tests. I think I have good coverage over the few functions I implemented. I expanded the integration test `test_train_with_different_data_and_training_device` and added a new integration test `test_embedding_nets_integration_training_device` that tackles the use case that was failing for me. It checks that everything runs even with wrong device assignments and that the users is warned when data or modules are moved automatically.

LouisRouillard commented:
I sadly have conflicts, I'll try to rebase over main and pray it's not too bad ^^'

michaeldeistler commented:
Watch out for `sbi/utils/plot.py` -- the file got deleted along the way

michaeldeistler commented:
Great! Is this ready for review?

michaeldeistler commented:
Ah just saw discord...I'll review it now

michaeldeistler commented on sbi/neural_nets/classifier.py:
```diff
@@ -114,6 +115,15 @@ def build_linear_classifier(
+        "call the build function using data from a common device."
+    )
+    check_embedding_net_device(embedding_net=embedding_net_x, datum=batch_y)
+    check_embedding_net_device(embedding_net=embedding_net_y, datum=batch_y)
```
this is great!

michaeldeistler commented on sbi/utils/torchutils.py:
```diff
@@ -21,34 +21,38 @@ def process_device(device: str) -> str:
+        )
+        current_gpu_index = torch.cuda.current_device()
+        if device == "cuda":
+            return f"cuda:{current_gpu_index}"
```
awesome!

michaeldeistler reviewed (approved):
This is great! Thanks a lot for looking into this and understanding and fixing the issue. Good to go from my side!

LouisRouillard commented:
All test in `tests/inference_on_device_test.py` and `pytest tests/user_input_checks_test.py` including `slow` and `gpu` run locally !

janfb commented on sbi/neural_nets/classifier.py:
```diff
@@ -114,6 +115,15 @@ def build_linear_classifier(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
copy paste typo: `y` vs. `x`

LouisRouillard replied:
Good call ! Will update all.

janfb commented on sbi/neural_nets/classifier.py:
```diff
@@ -161,6 +171,15 @@ def build_mlp_classifier(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
typo: `y` vs `x`

janfb commented on sbi/neural_nets/classifier.py:
```diff
@@ -216,6 +235,15 @@ def build_resnet_classifier(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
typo: `y` vs `x`

janfb commented on sbi/neural_nets/flow.py:
```diff
@@ -52,6 +53,13 @@ def build_made(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
typo: `y` vs `x`

janfb commented on sbi/neural_nets/flow.py:
```diff
@@ -123,6 +131,13 @@ def build_maf(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
typo: `y` vs `x`

janfb commented on sbi/neural_nets/flow.py:
```diff
@@ -205,6 +220,13 @@ def build_nsf(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
typo: `y` vs `x`

janfb commented on sbi/neural_nets/mdn.py:
```diff
@@ -43,6 +45,13 @@ def build_mdn(
+    assert batch_x.device == batch_y.device, (
+        "Mismatch in fed data's device: "
+        f"batch_x has device '{batch_x.device}' whereas "
+        f"batch_x has device '{batch_x.device}'. Please "
```
typo: `y` vs `x`

janfb replied:
given we have >5 of those asserts you could write a function that checks two tensors for their device and prints a (slightly more general) error message?

LouisRouillard replied:
Will do !

janfb commented on sbi/utils/torchutils.py:
```diff
@@ -21,34 +21,38 @@ def process_device(device: str) -> str:
 
 
 def check_if_prior_on_device(device, prior: Optional[Any] = None):
     if prior is not None:
```
general suggestion: we could have a 
```
if prior is None: 
    pass
else: 
    ...
```
to improve readability?

LouisRouillard replied:
Will change it thanks :)

janfb commented on sbi/utils/user_input_checks.py:
```diff
@@ -395,6 +395,27 @@ def check_prior_support(prior):
         )
 
 
+def check_embedding_net_device(embedding_net: nn.Module, datum: torch.Tensor) -> None:
```
please add a short docstring.

LouisRouillard replied:
Will do !

janfb commented on sbi/utils/user_input_checks.py:
```diff
@@ -629,13 +650,20 @@ def validate_theta_and_x(
-        logging.warning(
-            f"""Simulations are on {simulations_device} but training device is
-            set to {training_device}, moving data to device to {training_device}."""
+    if str(x.device) != training_device:
```
👍

janfb reviewed (changes requested):
great code, great tests, thanks a lot! 
I added small comments on types and some refactoring. This is good to go in once they are addressed.

LouisRouillard reviewed (commented):
Thanks @janfb for the comments !

janfb reviewed (approved):
Great, thanks a lot! Go to go in once CI is passing
